stores.js

```diff
@@ -0,0 +1,179 @@
+    if (o == null) return o
+    return MutableStore.encode(o)
+  },
+  decode (o) {
+    if (o == null) return o
+    return MutableStore.decode(o)
+  }
+}
+
+const PUT_VALUE_MAX_SIZE = 1000
+
+const immutable = (store) => ({
+  get (key, cb) {
+    if (typeof cb !== 'function') throw Error('Callback is required')
+    if (!Buffer.isBuffer(key)) throw Error('Key must be a buffer')
+    const hexKey = key.toString('hex')
+    // if the querying node already has the immutable value
+    // then there's no need to query the dht
+    const value = store.get(hexKey)
+
+    if (value) return process.nextTick(cb, null, value)
+
+    const queryStream = this.query('immutable-store', key)
+    let found = false
+    queryStream.on('data', (result) => {
+      const check = Buffer.alloc(32)
+      hash(check, result.value)
+      if (Buffer.compare(check, key) === 0) {
+        found = true
+        cb(null, result.value)
+        queryStream.destroy()
+      } // silently ignore bad values
+    })
+
+    finished(queryStream, (err) => {
+      if (err) {
+        cb(err)
+        return
+      }
+      if (found === false) {
+        cb(Error('Not Found'))
+      }
+    })
+  },
+  put (value, cb) {
+    if (typeof cb !== 'function') throw Error('Callback is required')
+    if (!Buffer.isBuffer(value)) throw Error('Value must be a buffer')
+    if (value.length > PUT_VALUE_MAX_SIZE) { throw Error(`Value size must be <= ${PUT_VALUE_MAX_SIZE}`) }
+    const key = Buffer.alloc(32)
+    hash(key, value)
+    // set locally for easy cached retrieval
+    store.set(key.toString('hex'), value)
+    // send to the dht
+    this.update('immutable-store', key, value, (err) => {
+      if (err) {
+        cb(err)
+        return
+      }
+      cb(null, key)
+    })
+  },
+  command: {
+    update ({ target, value }, cb) {
+      if (value == null) {
+        cb(null)
+        return
+      }
+      const key = target.toString('hex')
+      store.set(key, value)
+      cb(null)
+    },
+    query ({ target }, cb) {
+      cb(null, store.get(target.toString('hex')))
+    }
+  }
+})
+
+const mutable = (store) => ({
+  keypair () {
+    const pk = Buffer.alloc(pkSize)
+    const sk = Buffer.alloc(skSize)
+    createKeypair(pk, sk)
+    return { pk, sk }
+  },
+  get (key, opts, cb = opts) {
+    if (typeof cb !== 'function') throw Error('Callback is required')
+    if (typeof opts !== 'object') throw Error('Options are required')
+    const { salt, seq } = opts
+    if (typeof seq !== 'number') throw Error('seq is a required option')
+    if (salt && !Buffer.isBuffer(key)) throw Error('salt must be a buffer')
+    const queryStream = this.query('mutable-store', key)
+    let found = false
+    queryStream.on('data', (result) => {
+      if (result.value === null) return
+      const { key, value, sig, seq: _seq, salt } = result.value
+      if (_seq === seq && verify(sig, value, key)) {
+        found = true
+        cb(null, value, { key, sig, seq, salt })
```

Comment:
We prob need an option to continue scanning to see if a newer one comes along (can be v2 feature)

Reply:
so for that case could we just allow no callback to be passed, and then return a stream? (which would be a transform stream that query stream is connected to?

Reply:
great idea, stream should contain every hit >= seq

Reply:
or in this case when `_seq <= seq` right?

Reply:
yes